File: frontend/src/domain/library.ts
/** Managed local Skill/Rules library types (Phase 17). */

export type AssetKind = "skill" | "rules";

export type LibraryValidationIssue = {
  code: string;
  message: string;
  nodeId?: string | null;
  edgeId?: string | null;
};

export type LibraryAssetManifest = {
  id: string;
  kind: AssetKind;
  name: string;
  description: string;
  originalFilename: string;
  importedAt: string;
  frontmatter: Record<string, unknown>;
  body: string;
};

export type LibraryAsset = {
  manifest: LibraryAssetManifest;
  originalContent: string;
};

export type LibraryAssetSummary = {
  id: string;
  kind: AssetKind;
  name: string;
  description: string;
  originalFilename: string;
  importedAt: string;
};

export type LibraryPreviewRequest = {
  filename: string;
  content: string;
  kind?: AssetKind | null;
};

export type LibraryPreviewResponse = {
  ok: boolean;
  kind?: AssetKind | null;
  name?: string | null;
  description?: string | null;
  frontmatter?: Record<string, unknown> | null;
  body?: string | null;
  originalContent?: string | null;
  originalFilename?: string | null;
  errors: LibraryValidationIssue[];
};

export type LibraryImportRequest = {
  filename: string;
  content: string;
  kind?: AssetKind | null;
};

export type LibraryImportResponse = {
  ok: boolean;
  asset?: LibraryAsset | null;
  errors: LibraryValidationIssue[];
};

export type LibraryListResponse = {
  assets: LibraryAssetSummary[];
};

export type LibraryBatchImportResponse = {
  results: LibraryImportResponse[];
  importedCount: number;
  failedCount: number;
};

/** Pending file awaiting preview/confirm in the UI. */
export type PendingLibraryFile = {
  filename: string;
  content: string;
  kindHint?: AssetKind | null;
};

File: backend/src/mitos_api/domain/library.py
"""Managed local asset library models (Phase 17)."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from mitos_api.domain.workflow import ValidationIssue


class AssetKind(str, Enum):
    SKILL = "skill"
    RULES = "rules"
    KNOWLEDGE_BASE = "knowledgeBase"


class LibraryAssetManifest(BaseModel):
    """Normalized manifest stored alongside the original file bytes."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    kind: AssetKind
    name: str = Field(min_length=1)
    description: str = ""
    originalFilename: str = Field(min_length=1)
    importedAt: str
    frontmatter: dict[str, Any] = Field(default_factory=dict)
    body: str = ""


class LibraryAsset(BaseModel):
    """Full library entry: original content + normalized manifest."""

    model_config = ConfigDict(extra="forbid")

    manifest: LibraryAssetManifest
    originalContent: str


class LibraryAssetSummary(BaseModel):
    """List view of a library asset (no body/original)."""

    model_config = ConfigDict(extra="forbid")

    id: str
    kind: AssetKind
    name: str
    description: str
    originalFilename: str
    importedAt: str


class LibraryPreviewRequest(BaseModel):
    """Preview a Markdown skill/rules file before confirming import."""

    model_config = ConfigDict(extra="forbid")

    filename: str = Field(min_length=1)
    content: str
    kind: AssetKind | None = None


class LibraryPreviewResponse(BaseModel):
    """Preview result — does not write to the managed library."""

    model_config = ConfigDict(extra="forbid")

    ok: bool
    kind: AssetKind | None = None
    name: str | None = None
    description: str | None = None
    frontmatter: dict[str, Any] | None = None
    body: str | None = None
    originalContent: str | None = None
    originalFilename: str | None = None
    errors: list[ValidationIssue] = Field(default_factory=list)


class LibraryImportRequest(BaseModel):
    """Confirm import of a previously previewed Markdown file."""

    model_config = ConfigDict(extra="forbid")

    filename: str = Field(min_length=1)
    content: str
    kind: AssetKind | None = None


class LibraryImportResponse(BaseModel):
    """Result of confirming an import into the managed library."""

    model_config = ConfigDict(extra="forbid")

    ok: bool
    asset: LibraryAsset | None = None
    errors: list[ValidationIssue] = Field(default_factory=list)


class LibraryListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    assets: list[LibraryAssetSummary] = Field(default_factory=list)


class LibraryBatchImportRequest(BaseModel):
    """Import several files in one request (gate: one Skill + multiple Rules)."""

    model_config = ConfigDict(extra="forbid")

    files: list[LibraryImportRequest] = Field(min_length=1)


class LibraryBatchImportResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    results: list[LibraryImportResponse]
    importedCount: int
    failedCount: int
